File: Swissmap.R
library(tidygeocoder)
library(leaflet)
library(leaflet.extras)
library(htmltools)
library(plotly)
library(shiny)
library(shinydashboard)
library(rpivotTable)
library(rgdal)
library(rgeos)
library(haven)
library(sjlabelled)
library(dplyr)
library(tidyverse)
library(maptools)
library(classInt)
library(scales)


#load shipefiles
shapefile_path <- "W:/07_Forschung/07_MACH Consumer/28_Consumer 2025/Organisation/Alternativ zu Tableau/Gemeinden.shp"
layers <- ogrListLayers(dsn = shapefile_path)
all_layers <- list()
for (layer_name in layers) {
  layer_data <- readOGR(dsn = shapefile_path, layer = layer_name)
  all_layers[[layer_name]] <- layer_data
}
newdf <- merge(all_layers[[1]], all_layers[[2]], by = "BFS_NUMMER")
for (i in 3:length(all_layers)) {
  newdf <- merge(newdf, all_layers[[i]], by = "BFS_NUMMER")
}
newdf <- newdf[newdf$OBJEKTART == "Gemeindegebiet", ]
newdf <- spTransform(newdf, CRS("+init=epsg:4326"))

# Import data
faux <- read.csv(...)

#Import Bfs-Number and Zip-Code
bfs <- read.csv("W:/07_Forschung/07_MACH Consumer/28_Consumer 2025/Organisation/Alternativ zu Tableau/Datensaetze/bfsnummern.csv")

# Merge data
merged <- merge(faux, bfs, by.x = "zip", by.y = "PLZ", all.x = TRUE)
merged <- merged[!duplicated(merged$id), ]
merged2 <- merged %>% group_by(WOBFS) %>% summarize(counts = n())
merged_map <- merge(newdf, merged2, by.x = "BFS_NUMMER", by.y = "WOBFS", all.x = TRUE)
merged_map@data$counts[is.na(merged_map@data$counts)] <- 0

# Create color palette
bins <- c(0, 1, 2, 5, 10, Inf)
pal <- colorBin("plasma", domain = merged_map@data[["counts"]], bins = bins)

# Create leaflet map
swiss2 <- leaflet(merged_map) %>%
  addTiles() %>%
  addPolygons(
    fillColor = ~pal(merged_map@data[["counts"]]),
    weight = 2,
    opacity = 1,
    color = "white",
    dashArray = "1",
    fillOpacity = 1,
    popup = paste("Gemeinde: ", merged_map@data[["NAME"]], "<br>",
                  "Anzahl Interviews: ", merged_map@data[["counts"]])
  ) %>%
  addLegend(pal = pal, values = merged_map@data[["counts"]], labels = merged_map@data[["counts"]], title = "Counts", position = "bottomright")

File: Swissmap.py
# Data Import and Visualization
# Import necessary packages
import pyreadstat
import pandas as pd
import geopandas as gpd
import networkx as nx
import matplotlib.pyplot as plt
import folium
from keplergl import KeplerGl
import descartes
from pyproj import Proj, transform
from functools import partial
from shapely.geometry import Polygon
import shapely

# Load GeoDataFrame and filter relevant data
# Visualize the first few rows of the GeoDataFrame
gemeinde = gpd.read_file(r"~/Gemeinden.shp")
newdf = gemeinde.query('OBJEKTART == "Gemeindegebiet"')
print(gemeinde.head())

# Transform GeoDataFrame to WGS84 coordinate reference system
nc_counties = newdf.to_crs("epsg:4326")

# Read and preprocess csv datasets
df1 = pd.read_csv(r'~/faux.csv')
df2 = pd.read_csv(r'~/bfsnummern.csv')

# Filter and transform relevant variables
df2 = df2[["PLZ", "WOBFS", "Region"]]
df1['zip'] = df1['zip'].astype("Int64")

# Merge datasets and handle duplicates
merged = pd.merge(df1, df2, left_on='zip', right_on='PLZ', how='inner')
merged = merged.drop_duplicates(subset=['MACH_ID'])

# Group by postal code and BFS number, then merge with GeoDataFrame
merged2 = merged.groupby(['PLZ', 'WOBFS']).size().reset_index(name='counts')
merged_map = pd.merge(nc_counties, merged2, left_on='BFS_NUMMER', right_on='WOBFS', how='left')

# Aggregate counts for each area
interarea = merged_map.groupby(['NAME'])['counts'].sum().reset_index(name='Anzahl')

# Merge with GeoDataFrame and handle missing values
dfgreen = merged_map.merge(right=interarea, how='inner', left_on="NAME", right_on="NAME")
dfgreen = dfgreen.fillna(0)

# Create choropleth map using KeplerGl
m = dfgreen.explore(
    column="Anzahl",          # Choropleth based on "Anzahl" column
    tooltip=["NAME", "PLZ", "Anzahl"],  # Show information on hover
    popup=True,               # Show all values on click
    tiles="CartoDB positron", # Use "CartoDB positron" tiles
    cmap="Set1",              # Use "Set1" matplotlib colormap
    style_kwds=dict(color="black"),  # Use black outline
)

# Save the map as an HTML file
m.save('~/swissmap.html')
